cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -123,15 +144,42 @@ public void initialize(SparkHttpServiceContext context) throws Exception {
       readLimit = readLimitString == null ? 10000 : Integer.valueOf(readLimitString);
     }
 
+    /**
+     * Read the secure key file to read the security key bytes and initialize Cipher using AES and the mode
+     * encryption/decryption is based on the parameter cipherMode.
+     *
+     * @param cipherMode either encrypt or decrypt mode
+     * @return Initialized Cipher
+     * @throws FileNotFoundException If the key file doesn't exist
+     * @throws IOException If there are any exception while reading the file
+     * @throws NoSuchAlgorithmException if encryption algorithm is not found
+     * @throws NoSuchPaddingException could be thrown while getting cipher instance for encryption algorithm
+     * @throws InvalidKeyException could be thrown while getting cipher instance for encryption algorithm
+     */
+    private Cipher initializeCipher(int cipherMode) throws FileNotFoundException, IOException, NoSuchPaddingException,
+      NoSuchAlgorithmException, InvalidKeyException {
+      Location keyLocation =
+        getDatasetBaseLocation(ReportGenerationApp.REPORT_FILESET).append(SparkPersistRunRecordMain.KEY_FILE_NAME);
+      if (!keyLocation.exists()) {
+        throw new FileNotFoundException("Security Key file doesn't exist, cannot share reports");
```

Comment:
`IllegalStateException` may make more sense here since the state of CDAP is not setup correctly.